vllm/core/scheduler_v2.py

```diff
@@ -0,0 +1,370 @@
+            scheduled_resumed_reqs=resumed_reqs_data,
+            scheduled_running_reqs=running_reqs_data,
+            num_scheduled_tokens=num_scheduled_tokens,
+            total_num_scheduled_tokens=total_num_scheduled_tokens,
+            preempted_req_ids=preempted_req_ids,
+            finished_req_ids=self.finished_req_ids,
+            aborted_req_ids=self.aborted_req_ids,
+        )
+
+        self.finished_req_ids = set()
+        self.aborted_req_ids = set()
+        return scheduler_output
+
+    def update_from_output(
+        self,
+        scheduler_output: "SchedulerOutput",
+        model_runner_output: "ModelRunnerOutput",
+    ) -> List[Request]:
+        sampled_token_ids = model_runner_output.sampled_token_ids_cpu.tolist()
+        num_scheduled_tokens = scheduler_output.num_scheduled_tokens
+        new_running: List[Request] = []
+        finished_reqs: List[Request] = []
+        for request in self.running:
+            req_id = request.request_id
+            # TODO: Consider speculative decoding.
+            request.num_computed_tokens += num_scheduled_tokens[req_id]
+            if request.num_computed_tokens >= request.num_prompt_tokens:
+                req_index = model_runner_output.req_id_to_index[req_id]
+                token_id = sampled_token_ids[req_index]
+                request.output_token_ids.append(token_id)
+                # TODO: Update the KV cache manager for prefix caching.
+
+                if (request.num_tokens >= self.max_model_len
+                        or request.num_output_tokens >= request.max_tokens):
+                    request.status = RequestStatus.FINISHED_LENGTH_CAPPED
+                    self.finished_req_ids.add(req_id)
+                    finished_reqs.append(request)
+                    self._free_request(request)
+                    continue
+            new_running.append(request)
+        self.running = new_running
+        return finished_reqs
+
+    def add_request(self, request: Request) -> None:
+        self.waiting.append(request)
+
+    def abort_requests(self, request_ids: Union[str, Iterable[str]]) -> None:
+        if isinstance(request_ids, str):
+            request_ids = (request_ids, )
+        request_ids = set(request_ids)
+
+        # TODO: Optimize this.
+        for queue in [self.waiting, self.running]:
+            aborted_reqs: List[Request] = []
+            for request in queue:
+                if not request_ids:
+                    break
+                if request.request_id in request_ids:
+                    request.status = RequestStatus.FINISHED_ABORTED
+                    aborted_reqs.append(request)
+                    request_ids.remove(request.request_id)
+
+            for request in aborted_reqs:
+                queue.remove(request)
+                self.aborted_req_ids.add(request.request_id)
+                self._free_request(request)
+
+    def stop_requests(self, request_ids: Union[str, Iterable[str]]) -> None:
```

Comment:
What are the differences between stop and abort? Can we merge the two functions?